src/tls.rs

```diff
@@ -1,6 +1,234 @@
-//! TLS-related helper types
-use core::ffi::{c_char, CStr};
-use core::fmt::Debug;
+//! Type safe abstraction for esp-tls
+
+use std::{
+    ffi::CStr,
+    fmt::{Debug, Display},
+    io::{self, Read, Write},
+};
+
+pub struct EspTls {
+    raw: *mut esp_idf_sys::esp_tls,
+}
+
+impl EspTls {
+    pub fn new(host: &str, port: i32, cfg: Config) -> Result<Self> {
+        // TODO: where to put async version? seperate struct?
+        let tls = unsafe { esp_idf_sys::esp_tls_init() };
+        if tls == std::ptr::null_mut() {
+            return Err(Error::BadAlloc);
+        }
+        let mut rcfg: esp_idf_sys::esp_tls_cfg = unsafe { std::mem::zeroed() };
+
+        if let Some(ca_cert) = cfg.ca_cert {
+            rcfg.__bindgen_anon_1.cacert_buf = ca_cert.data().as_ptr();
+            rcfg.__bindgen_anon_2.cacert_bytes = ca_cert.data().len() as u32;
+        }
+
+        if let Some(client_cert) = cfg.client_cert {
+            rcfg.__bindgen_anon_3.clientcert_buf = client_cert.data().as_ptr();
+            rcfg.__bindgen_anon_4.clientcert_bytes = client_cert.data().len() as u32;
+        }
+
+        if let Some(client_key) = cfg.client_key {
+            rcfg.__bindgen_anon_5.clientkey_buf = client_key.data().as_ptr();
+            rcfg.__bindgen_anon_6.clientkey_bytes = client_key.data().len() as u32;
+        }
+
+        if let Some(ckp) = cfg.client_key_password {
+            // TODO: maybe this is not actually required to be NUL-terminated
+            rcfg.clientkey_password = ckp.as_ptr() as *const u8;
+            rcfg.clientkey_password_len = ckp.to_bytes().len() as u32;
+        }
+
+        let mut alpn_protos: Vec<*const i8>;
```

Comment:
Can we get by with `heapless::Vec` or `[*const i8; N]` array? Requiring `alloc` just for this does not seem reasonable.